src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,321 @@
+        /// <returns>ComplexTime advanced by provided gate time</returns>
+        internal ComplexTime AdvanceBy(double time) {
+            if (time == 0.0) {
+                return new ComplexTime(DepthTime, DependencyTime + 1);
+            }
+            return new ComplexTime(DepthTime + time);
+        }
+
+        /// <summary>
+        /// Subtracts argument ComplexTime from "this" ComplexTime. Returns the result.
+        /// Depth times are assumed to be precise  for comparison.
+        /// </summary>
+        /// <param name="time">Time to subtract from this object.</param>
+        /// <returns>Result of subtraction.</returns>
+        internal ComplexTime Subtract(ComplexTime time) {
+            if (DepthTime < time.DepthTime) {
+                throw new ArgumentException("Result of ComplexTime subtraction is negative.");
+            }
+            if (DepthTime == time.DepthTime) {
+                return new ComplexTime(0, DependencyTime - time.DependencyTime);
+            }
+            return new ComplexTime(DepthTime - time.DepthTime, 0);
+        }
+
+        /// <summary>
+        /// Compares two complex times. First DepthTime is compared, then DependencyTime is compared.
+        /// </summary>
+        /// <returns>0 when a = b, -1 when a&lt; b, and 1 when a &gt; b</returns>
+        internal static int Compare(ComplexTime a, ComplexTime b) {
+            int result = a.DepthTime.CompareTo(b.DepthTime);
+            if (result != 0) {
+                return result;
+            }
+            return a.DependencyTime.CompareTo(b.DependencyTime);
+        }
+
+        /// <summary>
+        /// Finds smallest of the two ComplexTime arguments.
+        /// </summary>
+        /// <returns>Smallest of the two arguments according to comparison.</returns>
+        internal static ComplexTime Min(ComplexTime a, ComplexTime b) {
+            return Compare(a, b) < 0 ? a : b;
+        }
+
+        /// <summary>
+        /// Finds largest of the two ComplexTime arguments.
+        /// </summary>
+        /// <returns>Largest of the two arguments according to comparison.</returns>
+        internal static ComplexTime Max(ComplexTime a, ComplexTime b) {
+            return Compare(a, b) > 0 ? a : b;
+        }
+
+        /// <summary>
+        /// Minimal value of ComplexTime.
+        /// </summary>
+        internal static ComplexTime MinValue { get; } = new ComplexTime(double.MinValue, long.MinValue);
+
+        /// <summary>
+        /// Maximum value of ComplexTime.
+        /// </summary>
+        internal static ComplexTime MaxValue { get; } = new ComplexTime(double.MaxValue, long.MaxValue);
+
+        /// <summary>
+        /// Zero ComplexTime.
+        /// </summary>
+        internal static ComplexTime Zero { get; } = new ComplexTime(0.0, 0);
+    }
+
+    /// <summary>
+    /// Stores qubits sorted by corresponding times. Instantiated to store interval start points and end points.
```

Comment:
what interval?

Reply:
Updated wording will be in the next push.